rfcs/text/0000-simple-updates.md

```diff
@@ -0,0 +1,159 @@
+
+
+* **Step 3 - Wrangler edits the spreadsheet with their change**
+
+The wrangler edits the spreadsheet. They can change any metadata cell apart from the ones containing the UUIDs that ingest added. They can add and remove entries from tabs that contain array/module data. For this solution, they must not add or remove whole entities (e.g. biomaterials, protocols) or change the linking between entities. Both these are instances of experimental design representation changes and are out of scope.
+
+* **Step 4 - Wrangler submits spreadsheet**
+
+The wrangler will submit a spreadsheet containing an update through a “Submit update” button or similar on the submission page. For each submission, we will allow only one update to be submitted and processed at a time. For this solution, the wrangler can only submit an updated spreadsheet that validates against the current schemas. 
+
+* **Step 5 - Ingest shows the diff to the wrangler for approval**
+
+The wrangler will be shown the difference between their updates and the existing submission in the ingest UI. They will also be shown any validation errors. If there are no validation errors then they will have the option either to approve the update, in which case ingest will start submitting it to the datastore, or cancel it, in which case the update will be removed from ingest and they can return to step 3. If there are validation errors then the wrangler can only cancel the update and submit a new update spreadsheet.
+
+* **Step 6 - Ingest uploads new versions of metadata files and bundles to the datastore**
+
+For this solution, ingest will calculate the bundle updates and submit these to the datastore. This represents a risk to the scalability of this solution. If performance problems emerge we will need to consider solutions in a future RFC.
+
+### DCP Flow of Updates
+
+![](../images/0000-simple-updates-dcp-flow.png)
+
+Once the bundles are updated (1 & 2), the datastore will send update notifications to all listeners (3). The query service, data browser and matrix service **MUST** all process these notifications to provide the updated metadata to consumers as required. The data processing pipeline execution service **MUST NOT** submit new or updated analyses to ingest for assays that have previously had analysis submitted. This solution requires the data processing pipeline execution service to capture and check if notification has already triggered a pipeline run and stop the run if it has.
+
+For simple updates, ingest will update secondary bundles with primary data (copy forward) at the same time that it updates the primary bundles. It does not rely on a signal from the analysis component. This should change in future iterations of the update process.
```

Comment:
In the future, when analysis can process updates:

a) If copy-forward is still in place, analysis needs to signal that they decided not to re-run analysis in response to an update, at which point ingest needs to update the primary data in all the analysis bundles.
b) If analysis does run in response to an update, ingest needs to know:
  i) if the analysis is entirely new (not an update to an existing analysis) or
  ii) is the analysis updates an existing analysis.
Ingest cannot make that determination itself.
c) If an analysis is an update, ingest needs to know which files in the new analysis are updates to files in the existing analysis, so it can tell that datastore that these are file updates rather than entirely new files. Simplification with cast iron rules might be possible, e.g. that any file in the new analysis with the same path is an update to the file with the same path in the previous analysis, that any file in the old analysis but not in the new should be dropped. If we can't adopt these cast iron rules then the submitter (analysis) will need to specify - ingest cannot make those determinations.